File: internal/objectType/objectType.go
package objectType

import "strings"

type Type string

const (
	Table                  Type = "Table"
	TableExtention         Type = "TableExtension"
	Page                   Type = "Page"
	PageExtention          Type = "PageExtension"
	Report                 Type = "Report"
	ReportExtention        Type = "ReportExtension"
	Enum                   Type = "Enum"
	EnumExtention          Type = "EnumExtension"
	PermissionSet          Type = "PermissionSet"
	PermisisonSetExtention Type = "PermissionSetExtension"
	Codeunit               Type = "Codeunit"
	XMLPort                Type = "XMLPort"
	Query                  Type = "Query"
	MenuSuite              Type = "MenuSuite"
	Unknown                Type = "Unknown"
)

func MapObjectType(objectType string) Type {
	switch strings.ToLower(objectType) {
	case strings.ToLower(string(Table)):
		return Table
	case strings.ToLower(string(TableExtention)):
		return TableExtention
	case strings.ToLower(string(Page)):
		return Page
	case strings.ToLower(string(PageExtention)):
		return PageExtention
	case strings.ToLower(string(Report)):
		return Report
	case strings.ToLower(string(ReportExtention)):
		return ReportExtention
	case strings.ToLower(string(Enum)):
		return Enum
	case strings.ToLower(string(EnumExtention)):
		return EnumExtention
	case strings.ToLower(string(PermissionSet)):
		return PermissionSet
	case strings.ToLower(string(PermisisonSetExtention)):
		return PermisisonSetExtention
	case strings.ToLower(string(Codeunit)):
		return Codeunit
	case strings.ToLower(string(XMLPort)):
		return XMLPort
	case strings.ToLower(string(Query)):
		return Query
	case strings.ToLower(string(MenuSuite)):
		return MenuSuite
	default:
		return Unknown
	}
}

func GetObjectTypes() []Type {
	return []Type{
		Table,
		TableExtention,
		Page,
		PageExtention,
		Report,
		ReportExtention,
		Enum,
		EnumExtention,
		PermissionSet,
		PermisisonSetExtention,
		Codeunit,
		XMLPort,
		Query,
		MenuSuite,
	}
}

File: src/object_type.rs
use std::{
    fmt::{self, Display},
    str::FromStr,
};

#[derive(Debug, Hash, Eq, PartialEq)]
pub enum ObjectType {
    None,
    Codeunit,
    Table,
    TableExtension,
    Page,
    PageExtension,
    Report,
    ReportExtension,
    Enum,
    EnumExtension,
    PermissionSet,
    PermissionSetExtension,
    MenuSuite,
    Query,
    XMLPort,
}

impl FromStr for ObjectType {
    type Err = String;

    fn from_str(input: &str) -> Result<ObjectType, Self::Err> {
        match input {
            "Codeunit" => Ok(ObjectType::Codeunit),
            "Table" => Ok(ObjectType::Table),
            "TableExtension" => Ok(ObjectType::TableExtension),
            "Page" => Ok(ObjectType::Page),
            "PageExtension" => Ok(ObjectType::PageExtension),
            "Report" => Ok(ObjectType::Report),
            "ReportExtension" => Ok(ObjectType::ReportExtension),
            "Enum" => Ok(ObjectType::Enum),
            "EnumExtension" => Ok(ObjectType::EnumExtension),
            "PermissionSet" => Ok(ObjectType::PermissionSet),
            "PermissionSetExtension" => Ok(ObjectType::PermissionSetExtension),
            "MenuSuite" => Ok(ObjectType::MenuSuite),
            "Query" => Ok(ObjectType::Query),
            "XMLPort" => Ok(ObjectType::XMLPort),
            _ => Err("unknown object type".to_string()),
        }
    }
}

impl Display for ObjectType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                ObjectType::None => "None",
                ObjectType::Codeunit => "Codeunit",
                ObjectType::Table => "Table",
                ObjectType::TableExtension => "TableExtension",
                ObjectType::Page => "Page",
                ObjectType::PageExtension => "PageExtension",
                ObjectType::Report => "Report",
                ObjectType::ReportExtension => "ReportExtension",
                ObjectType::Enum => "Enum",
                ObjectType::EnumExtension => "EnumExtension",
                ObjectType::PermissionSet => "PermissionSet",
                ObjectType::PermissionSetExtension => "PermissionSetExtension",
                ObjectType::MenuSuite => "MenuSuite",
                ObjectType::Query => "Query",
                ObjectType::XMLPort => "XMLPort",
            }
        )
    }
}
